File: GolangSolutions/DP/M1235_规划兼职工作.go
package DP

import (
	"sort"
)

const N int = 50010

func jobScheduling(startTime []int, endTime []int, profit []int) int {
	n := len(startTime)
	jobs := make([][3]int, n)
	for i := range startTime {
		jobs[i] = [3]int{startTime[i], endTime[i], profit[i]}
	}
	sort.Slice(jobs, func(i, j int) bool { return jobs[i][1] < jobs[j][1] })

	f := make([]int, N)
	for i := 1; i <= n; i++ {
		last := sort.Search(i, func(j int) bool { return jobs[j][1] > jobs[i-1][0] })
		f[i] = max(f[i-1], f[last]+jobs[i-1][2])
	}

	return f[n]
}

File: CPPSolutions/LeetCode/DP/M1235_规划兼职工作.cpp
#include <bits\stdc++.h>

using namespace std;

const int N = 50010;

vector<int> f(N);

class Solution {
public:
    int jobScheduling(vector<int> &startTime, vector<int> &endTime, vector<int> &profit) {
        int n = startTime.size();
        vector<vector<int>> jobs(n);
        for (int i = 0; i < n; i++)
            jobs[i] = {startTime[i], endTime[i], profit[i]};
        sort(jobs.begin(), jobs.end(), [](const vector<int> &job1, const vector<int> &job2) -> bool {
            return job1[1] < job2[1];
        });

        for (int i = 1; i <= n; i++) {
            int last = upper_bound(jobs.begin(), jobs.begin() + i - 1, jobs[i - 1][0],
                                   [&](int st, const vector<int> &job) -> bool {
                                       return st < job[1];
                                   }) - jobs.begin();
            f[i] = max(f[i - 1], f[last] + jobs[i - 1][2]);
        }

        return f[n];
    }
};